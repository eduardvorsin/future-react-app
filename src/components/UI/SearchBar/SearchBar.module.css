.search-form {
	max-width: 300px;
}

.search-form__search-bar {
	/* .search-form__search-bar */
}

.search-bar {
	position: relative;

}

.search-bar__label {
	/* .search-bar__label styles */
}

.search-bar__input {
	width: 100%;
	padding: 7px 35px;
	border-width: 2px;
	border-style: solid;
	border-color: transparent;
	font-size: 18px;
	color: var(--text-color-primary);
	background-color: var(--secondary-color-200);
	border-radius: 4px;
	transition:
		border-color var(--transition-time) var(--transition-easing),
		background-color var(--transition-time) var(--transition-easing),
		color var(--transition-time) var(--transition-easing);
}

.search-bar__input::placeholder {
	color: var(--dark-color-200);
	transition: color var(--transition-time) var(--transition-easing);
}

.search-bar__input:hover:not([disabled]) {
	border-color: var(--brand-color-200);
	transition: border-color var(--transition-time) var(--transition-easing);
}

.search-bar__close {
	left: 0;
}

.search-bar__find {
	right: 0;
}

.search-bar__close,
.search-bar__find {
	position: absolute;
	top: 50%;
	transform: translateY(-50%);
}

.search-bar__close>svg,
.search-bar__find>svg {
	width: 25px;
	height: 25px;
}

@media(prefers-color-scheme:dark) {
	.search-bar__input::placeholder {
		color: var(--brand-color-100);
		transition: color var(--transition-time) var(--transition-easing);
	}
}