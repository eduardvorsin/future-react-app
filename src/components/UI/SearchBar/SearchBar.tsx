import React from 'react';
import Button from '../Button/Button';
import SearchIcon from '../../../assets/images/icons/find.svg';
import ClearIcon from '../../../assets/images/icons/clear.svg';
import classes from './SearchBar.module.css';

type BaseSearchBarProps = Partial<React.InputHTMLAttributes<HTMLInputElement>>;

interface SearchBarProps extends BaseSearchBarProps {
  className?: string,
  value: string | number,
  onClear: React.MouseEventHandler<HTMLButtonElement>,
  onChange: React.ChangeEventHandler<HTMLInputElement>,
  onSearch: () => void,
}

const SearchBar: React.FC<SearchBarProps> = ({
  className,
  value,
  onClear,
  onChange,
  onSearch,
  ...props
}) => {
  const SearchFormClasses = [
    className,
    classes['search-form'],
    classes['search-form__search-bar'],
  ].join(' ');

  const SearchBarClasses = [
    classes['search-bar'],
  ].join(' ');

  const keydownHandler: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
    if (e.key === 'Enter') {
      onSearch();
    }
  };

  const searchSubmitHandler: React.FormEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault();
    if (value.toString().length < 1) {
      return;
    }

    onSearch();
  };

  return (
    <form
      className={SearchFormClasses}
      onSubmit={searchSubmitHandler}
    >
      <div
        className={SearchBarClasses}
        role='search'
      >
        <label
          htmlFor='search-bar'
          className={`${classes['search-bar__label']} sr-only`}
        >
          Search bar
        </label>
        <input
          id='search-bar'
          className={classes['search-bar__input']}
          onChange={onChange}
          onKeyDown={keydownHandler}
          value={value}
          placeholder='найти книгу...'
          {...props}
        />
        <Button
          iconButton
          className={classes['search-bar__close']}
          onClick={onClear}
        >
          <ClearIcon />
        </Button>
        <Button
          iconButton
          className={classes['search-bar__find']}
          type='submit'
        >
          <SearchIcon />
        </Button>
      </div>
    </form>
  );
};

export default SearchBar;
