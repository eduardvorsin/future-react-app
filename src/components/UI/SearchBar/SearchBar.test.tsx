import React, { ChangeEvent, MouseEvent } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SearchBar from './SearchBar';
import i18n from '../../../localization/i18next';

describe('SearchBar tests', () => {
  it('is rendered correctly', () => {
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
        testId='test-searchbar'
      />,
    );

    expect(screen.getByTestId<HTMLFormElement>('test-searchbar')).toBeInTheDocument();
  });

  it('when you click on the clear button, the mock function should work', async () => {
    const user = userEvent.setup();
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
      />,
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /clear/i }));

    expect(mockClearHandler).toHaveBeenCalledTimes(1);
  });

  it('when you click on the find button, the mock function should work', async () => {
    const user = userEvent.setup();
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
      />,
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /find/i }));

    expect(mockFindHandler).toHaveBeenCalledTimes(1);
  });

  it('when entering a value in the search field, the mock function should be called', async () => {
    const user = userEvent.setup();
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
      />,
    );

    await user.type(screen.getByRole<HTMLInputElement>('textbox'), 'a');

    expect(mockChangeHandler).toHaveBeenCalledTimes(1);
  });

  it('is a basic snapshot', () => {
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
        testId='test-searchbar'
      />,
    );

    expect(screen.getByTestId<HTMLFormElement>('test-searchbar')).toMatchSnapshot();
  });
});

describe('SearchBar integration tests', () => {
  it('the component with the text in English is displayed correctly', () => {
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();
    const localizedLabelText = i18n.t('searchBar.label');
    const localizedClearButtonText = i18n.t('searchBar.clear');

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
        testId='test-searchbar'
      />,
    );

    expect(screen.getByLabelText<HTMLLabelElement>(localizedLabelText)).toBeInTheDocument();
    expect(screen.getByRole<HTMLButtonElement>('button', { name: localizedClearButtonText })).toBeInTheDocument();
  });

  it('the component with the text in Russian is displayed correctly', () => {
    i18n.changeLanguage('ru');
    const mockFindHandler = jest.fn<void, []>();
    const mockClearHandler = jest.fn<void, [e: MouseEvent]>();
    const mockChangeHandler = jest.fn<void, [e: ChangeEvent]>();
    const localizedLabelText = i18n.t('searchBar.label');
    const localizedClearButtonText = i18n.t('searchBar.clear');

    render(
      <SearchBar
        onClear={mockClearHandler}
        onSearch={mockFindHandler}
        onChange={mockChangeHandler}
        value='mock value'
        testId='test-searchbar'
      />,
    );

    expect(screen.getByLabelText<HTMLLabelElement>(localizedLabelText)).toBeInTheDocument();
    expect(screen.getByRole<HTMLButtonElement>('button', { name: localizedClearButtonText })).toBeInTheDocument();
  });
});
