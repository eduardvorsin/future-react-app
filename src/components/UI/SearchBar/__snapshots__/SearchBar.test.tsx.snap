// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar tests is a basic snapshot 1`] = `
<form
  class=" search-form search-form__search-bar"
  data-testid="test-searchbar"
>
  <div
    class="search-bar"
    role="search"
  >
    <label
      class="search-bar__label sr-only"
      for="search-bar"
    >
      Search Bar
    </label>
    <input
      class="search-bar__input"
      id="search-bar"
      placeholder="Find a book..."
      value="mock value"
    />
    <button
      class="button--icon-button search-bar__clear"
    >
      <svg />
      Clear
    </button>
    <button
      class="button--icon-button search-bar__find"
      type="submit"
    >
      <svg />
      To find
    </button>
  </div>
</form>
`;
